### ISSUE TITLE
ttl.tensor.reduce needs to be run twice to produce correct output

### ISSUE THREAD

**Issue by arakhmati** [tenstorrent/tt-metal#4384]
Running the following line once produces a tensor with incorrect values in some tiles:
```python
ttl_output_tensor = ttl.tensor.reduce(ttl_input_tensor, ttl.tensor.ReduceOpMath.SUM, reduce_op_dim, 1 / input_shape[-1])
```

Re-running it right after the second time produces the correct result

**Comment by arakhmati** [comment]
Running the following line once produces a tensor with incorrect values in some tiles:
 `ttl_output_tensor = ttl.tensor.reduce(ttl_input_tensor, ttl.tensor.ReduceOpMath.SUM, reduce_op_dim, 1 / input_shape[-1])`

Re-running it right after the second time produces the correct result

**Comment by tt-aho** [comment]
Is this for any reduce dim or a specific one?

**Comment by arakhmati** [comment]
> Is this for any reduce dim or a specific one?

It happens for both H and W dims. I haven't tried HW

**Comment by eyonland** [comment]
@jliangTT , we believe this to be very high priority as it could impact all models.

**Comment by jliangTT** [comment]
@umadevimcw , thanks for assigning @VirdhatchaniKN. We probably need to fast track this as this has been opened for a while.

**Comment by VirdhatchaniKN** [comment]
Hi @tt-aho , @arakhmati , @eyonland , @jliangTT 

I ran the test file `tests/ttnn/unit_tests/operations/test_mean.py` to work on the error, and as @arakhmati  mentioned, the reduce op needs to be run twice to get the correct result. But this assertion error is specific to dim -2 (height) only. Width seems to work fine.

- We tried implementing mean in TTDNN, so that we can export it to TTNN and update the file, but we faced issues with the result here (Assertion error).
- On debugging , It was found that reduce result did not match with the expected result. We implemented using `reduce`, `sum`, `mean_hw` , but we could still see no change.

I have discussed in detail about the code, and PCC value for each method for the same formula (mean = total/dimension) in this document - [Link](https://docs.google.com/document/d/1l9m6dQlQL2QN4MJxq12lRaA5UmKtBA04/edit?usp=sharing&ouid=111601744705016142724&rtpof=true&sd=true)

**Comment by umadevimcw** [comment]
@arakhmati @eyonland Any inputs on this to proceed further

**Comment by KalaivaniMCW** [comment]
@arakhmati @eyonland 
While using reduce op for mean calculation, the scaler value being passed is calculated with respect to the dimension being reduced. 
the scaler value for mean operation is calculated  as  **1 / size_of_dim_to_be_reduced** 
ex., for a tensor (1 , 2, 32 , 64), the scaler for different dimensions
dim= -1 , scaler = 1 / input_shape[-1] = 1/64
dim= -2 , scaler = 1 / input_shape[-2] = 1/32
dim= (-1, -2), scaler = 1 /(  input_shape[-1] * input_shape[-2]) = 1 / (32 * 64)
When testing with above scaler values for the different cases, the second reduce operation was not necessary
From this observation, the fix has been made in #7965
